src/ray/gcs/gcs_server/gcs_resource_report_poller.h

```diff
@@ -0,0 +1,133 @@
+  is _not_ thread safe (i.e. should not modify the resource report poller state).
+  5. (Polling thread) The RequestResourceReport continuation runs, scheduling the next
+  pull time.
+  6. (Polling thread) The next pull time occurs, and step 2 is repeated.
+
+  The node leaves the cluster.
+  7. Untrack the node. The next time the main polling procedure comes across the node, it
+  should be dropped from the system.
+   */
+
+ public:
+  GcsResourceReportPoller(
+      std::shared_ptr<GcsResourceManager> gcs_resource_manager,
+      std::shared_ptr<rpc::NodeManagerClientPool> raylet_client_pool,
+      std::function<void(const rpc::ResourcesData &)> handle_resource_report,
+      /* Default values should only be changed for testing. */
+      std::function<int64_t(void)> get_current_time_milli =
+          []() { return absl::GetCurrentTimeNanos() / (1000 * 1000); },
+      std::function<void(
+          const rpc::Address &, std::shared_ptr<rpc::NodeManagerClientPool> &,
+          std::function<void(const Status &, const rpc::RequestResourceReportReply &)>)>
+          request_report =
+              [](const rpc::Address &address,
+                 std::shared_ptr<rpc::NodeManagerClientPool> &raylet_client_pool,
+                 std::function<void(const Status &,
+                                    const rpc::RequestResourceReportReply &)>
+                     callback) {
+                auto raylet_client = raylet_client_pool->GetOrConnectByAddress(address);
+                raylet_client->RequestResourceReport(callback);
+              });
+
+  ~GcsResourceReportPoller();
+
+  /// Start a thread to poll for resource updates.
+  void Start();
+
+  /// Stop polling for resource updates.
+  void Stop();
+
+  /// Event handler when a new node joins the cluster.
+  void HandleNodeAdded(const std::shared_ptr<rpc::GcsNodeInfo> &node_info)
+      LOCKS_EXCLUDED(mutex_);
+
+  /// Event handler when a node leaves the cluster.
+  void HandleNodeRemoved(const std::shared_ptr<rpc::GcsNodeInfo> &node_info)
+      LOCKS_EXCLUDED(mutex_);
+
+ private:
+  // An asio service which does the polling work.
+  boost::asio::io_context polling_service_;
+  // The associated thread it runs on.
+  std::unique_ptr<std::thread> polling_thread_;
+  // Timer tick to check when we should do another poll.
+  PeriodicalRunner ticker_;
+
+  // The maximum number of pulls that can occur at once.
+  const uint64_t max_concurrent_pulls_;
+  // The number of ongoing pulls.
+  uint64_t inflight_pulls_;
+  // The resource manager which maintains GCS's view of the cluster's resource
+  // utilization.
+  std::shared_ptr<GcsResourceManager> gcs_resource_manager_;
+  // The shared, thread safe pool of raylet clients, which we use to minimize connections.
+  std::shared_ptr<rpc::NodeManagerClientPool> raylet_client_pool_;
+  // Handle receiving a resource report (e.g. update the resource manager).
+  std::function<void(const rpc::ResourcesData &)> handle_resource_report_;
```

Comment:
If this method accesses variables that aren't thread-safe to call from the polling_service_, can we add a comment to that effect?